x/cronos/keeper/evm.go

```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		return nil, err
+	}
+	if res.Failed() {
+		return nil, fmt.Errorf("call contract failed: %s, %s, %s", contract.Hex(), method, res.Ret)
+	}
+	return res.Ret, nil
+}
+
+// DeployModuleERC20 deploy an embed erc20 contract
+func (k Keeper) DeployModuleERC20(ctx sdk.Context, denom string) (common.Address, error) {
+	ctor, err := types.ModuleERC20Contract.ABI.Pack("", denom, uint8(0))
+	if err != nil {
+		return common.Address{}, err
+	}
+	data := append(types.ModuleERC20Contract.Bin, ctor...)
+
+	msg, res, err := k.CallEVM(ctx, nil, data, big.NewInt(0))
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	if res.Failed() {
+		return common.Address{}, fmt.Errorf("contract deploy failed: %s", res.Ret)
+	}
+	return crypto.CreateAddress(types.EVMModuleAddress, msg.Nonce()), nil
+}
+
+// SendCoinFromNativeToERC20 convert native token to erc20 token
+func (k Keeper) SendCoinFromNativeToERC20(ctx sdk.Context, sender common.Address, coin sdk.Coin, autoDeploy bool) error {
+	// TODO validate coin.Denom
+	var err error
+	// external contract is returned in preference to auto-deployed ones
+	contract, found := k.GetContractByDenom(ctx, coin.Denom)
+	if !found {
+		if !autoDeploy {
+			return errors.New("no contract found for the denom")
+		}
+		contract, err = k.DeployModuleERC20(ctx, coin.Denom)
+		if err != nil {
+			return err
+		}
+		k.SetAutoContractForDenom(ctx, coin.Denom, contract)
+	}
+	err = k.bankKeeper.SendCoins(ctx, sdk.AccAddress(sender.Bytes()), sdk.AccAddress(contract.Bytes()), sdk.NewCoins(coin))
+	if err != nil {
+		return err
+	}
+	_, err = k.CallModuleERC20(ctx, contract, "mint_by_native", sender, coin.Amount.BigInt())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SendCoinFromERC20ToNative convert erc20 token to native token
+func (k Keeper) SendCoinFromERC20ToNative(ctx sdk.Context, contract common.Address, receiver common.Address, coin sdk.Coin) error {
+	// TODO validate coin.Denom
+
+	// validate contract address
+	externalContract, externalFound := k.GetExternalContractByDenom(ctx, coin.Denom)
+	autoContract, autoFound := k.GetAutoContractByDenom(ctx, coin.Denom)
+	found := (externalFound && externalContract == contract) || (autoFound && autoContract == contract)
+	if !found {
+		return errors.New("the contract address is not mapped to the denom")
+	}
+
+	err := k.bankKeeper.SendCoins(ctx, sdk.AccAddress(contract.Bytes()), sdk.AccAddress(receiver.Bytes()), sdk.NewCoins(coin))
+	if err != nil {
+		return err
+	}
+
+	_, err = k.CallModuleERC20(ctx, contract, "burn_by_native", receiver, coin.Amount.BigInt())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ConvertCoinsFromNativeToERC20 convert native tokens to erc20 tokens
+func (k Keeper) ConvertCoinsFromNativeToERC20(ctx sdk.Context, sender common.Address, coins sdk.Coins, autoDeploy bool) error {
+	for _, coin := range coins {
+		if err := k.SendCoinFromNativeToERC20(ctx, sender, coin, autoDeploy); err != nil {
```

Comment:
rename also `SendCoinFromNativeToERC20` to `ConvertCoinFromNativeToERC20`

Reply:
done